srt-protocol/src/options/validation.rs

```diff
@@ -0,0 +1,268 @@
+use std::ops::Deref;
+
+pub trait Validation: Sized {
+    type Error;
+
+    fn is_valid(&self) -> Result<(), Self::Error>;
+
+    fn try_validate(self) -> Result<Valid<Self>, Self::Error> {
+        self.is_valid()?;
+        Ok(Valid(self))
+    }
+}
+
+pub trait CompositeValidation: Validation {
```

Comment:
Would be great to have some comments on what these traits semantically mean. Composite of what, exactly?

Reply:
So there is a design tensions here I was trying to explore.
1) I wanted the type system to enforce/encode validation for configuration
2) I wanted to use bare structs, i.e. all pub fields,  potentially easily serializable. So, no traditional builder pattern, which is  used often in Rust to solve this problem, yet at the cost of excessive boiler plate code in to tightly control mutable access to configuration data.
3)  I was unsure if a near completely flat options/configuration data structure that mirrored the reference implementation would be sufficient or if one with a little more conceptual organization/structure would be more desirable. I am exploring the later, and when I started to write validation for the options based off of the reference implementation, I ran into cases where I needed to do validation across multiple parts of the data in different structs, so I introduced a composite validation type. If we just went with a flat structure, this wouldn't be necessary. But if you like the shape of the "options" module's data, this is kind of essential to making it all work.

Also, I agree this needs more documentation, but I'm actually not sure yet if this structure is desirable/needed. I'm more curious for yours or other's feedback (@aegroto?).

Reply:
I'm more for keeping it simple and have a flat struct with public fields, although this would require for sure the configuration to be consumed by the socket's initialization. I'm not sure this will make the socket configurable dynamically even after the communication has started, have you thought about this scenario yet?

Reply:
When I read "simple" I reflexively reinterpret it as "easy" and "familiar", and I respect this perspective, ergonomics and familiarity are important. Though ease is often confounded with simple. Breaking apart pieces into coherent chunks could be more simple relative to the domain problem. Sometimes we are better off adapt our sense of ease/familiarity to the simplicity essential to the domain problem/model. See: https://www.infoq.com/presentations/Simple-Made-Easy/ To be clear, I'm not arguing this is one of those times, but thank you for indulging me soap box :)